machine-learning/jax-haiku-dask-dataframe-distributed-example.ipynb

```diff
@@ -0,0 +1,535 @@
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "ddf_validation[\"y_pred_with_training_CASE1\"] = ddf_validation.map_partitions(dask_predict_wrapper, params=params)\n",
+    "\n",
+    "(df_train.hvplot.scatter(x=\"x\", y=\"y\", label='Training data') * \n",
+    " ddf_validation.compute().hvplot.scatter(x=\"x\", y=\"y_pred_with_training_CASE1\", label='Predicted validation data')).opts(title=\"With training CASE 1\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### CASE 2 - Data-parallel training of deep learning models: Compute gradients in parallel."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "First let's reset the model parameters and optimizer state (so that we don't benefit from the training in Case 1)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "_, params = initialize_net_function()\n",
+    "opt_state = optimizer.init(params)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In Case 1, we cleverly load data only when it's needed (i.e. when we are ready to train that particular batch of data). However we are not benefiting from the parallel capabilities of Dask to train multiple batches at the same time.\n",
+    "\n",
+    "For this case, we compute the gradients of the loss function in parallel, and pull these back to the client where we update the model with the optimizer there."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "@jax.jit\n",
+    "def compute_grads(params: hk.Params, x: jnp.ndarray, y: jnp.ndarray) -> jnp.ndarray:\n",
+    "    grads = jax.grad(loss_function)(params, x, y)\n",
+    "    return grads\n",
+    "\n",
+    "def dask_compute_grads_one_partition_wrapper(ddf: dd.DataFrame, params: hk.Params) -> jnp.ndarray:\n",
+    "    scaled_x = jnp.array(ddf[[\"scaled_x\"]].values)\n",
+    "    y = jnp.array(ddf[[\"y\"]].values)\n",
+    "    grads = compute_grads(params, scaled_x, y)\n",
+    "    return grads"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The training loop now looks like:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "for epoch_number in range(200):  # num_epochs\n",
+    "    futures = []\n",
+    "    for ddf_one_partition in ddf_train.partitions:\n",
+    "        # Compute the gradients in parallel\n",
+    "        futures.append(client.submit(dask_compute_grads_one_partition_wrapper, ddf_one_partition, params))\n",
+    "    \n",
+    "    for future in as_completed(futures):\n",
+    "        # Bring the gradients back to the client, and update the model with the optimizer on the client\n",
+    "        grads = future.result()\n",
```

Comment:
See the `with_results=` keyword of as_completed to make this a bit faster.

Reply:
Thanks!